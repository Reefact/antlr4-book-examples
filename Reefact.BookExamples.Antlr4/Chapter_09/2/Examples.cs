#region Usings declarations

using Antlr4.Runtime;
using Antlr4.Runtime.Atn;

using ApprovalTests;
using ApprovalTests.Reporters;

using NFluent;

using Xunit;

#endregion

namespace Reefact.BookExamples.Antlr4.Chapter_09._2 {

    [UseReporter(typeof(BeyondCompareReporter), typeof(VisualStudioReporter))]
    public class Examples {

        [Fact]
        public void verbose_listener() {
            // Setup
            AntlrInputStream inputStream = AntlrInputStreamReader.Read("class T T { int ; }");
            SimpleGRun       grun        = SimpleGRun.Read(inputStream, new VerboseListener());
            // Exercise
            string output = grun.GetOutput();
            // Verify
            Approvals.Verify(output);
        }

        [Fact]
        public void underline_listener() {
            // Setup
            AntlrInputStream inputStream = AntlrInputStreamReader.Read("underline_listener.simple", 9, 2);
            SimpleGRun       grun        = SimpleGRun.Read(inputStream, new UnderlineListener());
            // Exercise
            string output = grun.GetOutput();
            // Verify
            Approvals.Verify(output);
        }

        [Fact]
        public void ambiguity_not_detected() {
            // Setup
            AntlrInputStream inputStream = AntlrInputStreamReader.Read("f();");
            AmbiguousGRun    grun        = AmbiguousGRun.Read(inputStream, new VerboseListener());
            // Exercise
            string output = grun.GetOutput();
            // Verify
            Check.That(output).IsEmpty();
        }

        [Fact]
        public void ambiguity_detected() {
            // Setup
            AntlrInputStream        inputStream             = AntlrInputStreamReader.Read("f();");
            DiagnosticErrorListener diagnosticErrorListener = new();
            VerboseListener         verboseListener         = new();
            var                     errorListeners          = new BaseErrorListener[] { verboseListener, diagnosticErrorListener };

            static void Options(Parser parser) {
                parser.Interpreter.PredictionMode = PredictionMode.LL_EXACT_AMBIG_DETECTION;
            }

            AmbiguousGRun grun = AmbiguousGRun.Read(inputStream, Options, errorListeners);

            // Exercise
            string output = grun.GetOutput();

            // Verify
            Approvals.Verify(output);
        }

    }

}