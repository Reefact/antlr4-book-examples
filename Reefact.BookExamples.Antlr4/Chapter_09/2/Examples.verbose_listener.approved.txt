rule stack: [prog -> classDef]
line 1:8 at [@2,8:8='T',<10>,1:8]: extraneous input 'T' expecting '{'
rule stack: [prog -> classDef -> member]
line 1:18 at [@5,18:18='}',<3>,1:18]: no viable alternative at input 'int}'
class T